src/mongoc/mongoc-topology-description.c

```diff
@@ -0,0 +1,996 @@
+   { /* UNKNOWN */
+      NULL, /* MONGOC_TOPOLOGY_UNKNOWN */
+      NULL, /* MONGOC_TOPOLOGY_SHARDED */
+      NULL, /* MONGOC_TOPOLOGY_RS_NO_PRIMARY */
+      _mongoc_topology_description_check_if_has_primary /* MONGOC_TOPOLOGY_RS_WITH_PRIMARY */
+   },
+   { /* STANDALONE */
+      _mongoc_topology_description_update_unknown_with_standalone,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_remove_and_check_primary
+   },
+   { /* MONGOS */
+      _mongoc_topology_description_set_topology_type_to_sharded,
+      NULL,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_remove_and_check_primary
+   },
+   { /* PRIMARY */
+      _mongoc_topology_description_update_rs_from_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_from_primary,
+      _mongoc_topology_description_update_rs_from_primary
+   },
+   { /* SECONDARY */
+      _mongoc_topology_description_transition_unknown_to_rs_no_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_without_primary,
+      _mongoc_topology_description_update_rs_with_primary_from_member
+   },
+   { /* ARBITER */
+      _mongoc_topology_description_transition_unknown_to_rs_no_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_without_primary,
+      _mongoc_topology_description_update_rs_with_primary_from_member
+   },
+   { /* RS_OTHER */
+      _mongoc_topology_description_transition_unknown_to_rs_no_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_without_primary,
+      _mongoc_topology_description_update_rs_with_primary_from_member
+   },
+   { /* RS_GHOST */
+      NULL,
+      _mongoc_topology_description_remove_server,
+      NULL,
+      _mongoc_topology_description_check_if_has_primary
+   }
+};
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * Handle an ismaster. This is called by the background SDAM process.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+bool
+mongoc_topology_description_handle_ismaster (
+   mongoc_topology_description_t *topology,
+   mongoc_server_description_t   *sd,
+   const bson_t                  *ismaster_response,
+   int64_t                        rtt_msec,
+   bson_error_t                  *error)
+{
+   bson_return_val_if_fail (topology, false);
+   bson_return_val_if_fail (sd, false);
+
+   mongoc_server_description_handle_ismaster (sd, ismaster_response, rtt_msec, error);
+
+   if (gSDAMTransitionTable[sd->type][topology->type]) {
+      gSDAMTransitionTable[sd->type][topology->type](topology, sd);
+   }
+
+   if (mongoc_topology_description_server_by_id (topology, sd->id)) {
```

Comment:
can we simplify this if to `return mongoc_topology_description_server_by_id (topology, sd->id) != NULL`

or have no return type for this function since nobody cares if it returns false.